SampleType-based HealthKit Query API

lukaskollmer opened the pull request:
# SampleType-based HealthKit Query API

## :recycle: Current situation & Problem
SpeziHealthKit offers a generic `SampleType` struct, which is used to work with `HKSampleType`s in a type-safe manner.
SpeziHealthKit currently also offers a `@HealthKitQuery` property wrapper which uses these `SampleType`s to allow easy, type-safe fetching of HealthKit data within SwiftUI views.

This PR adds some functions that implement simple, type-safe querying of HealthKit data outside of SwiftUI.
- `query(...) -> [Sample]`, which implements one-off querying of HealthKit samples
- `anchorQuery(...) -> (added: [Sample], deleted: [HKDeletedObject])`, which implements anchor queries of HealthKit data
- `continuousQuery(...) -> AsyncSequence`, which implements a long-running HealthKit query. this function returns an `AsyncSequence`, which will produce new elements every time HealthKit informs the query about new/deleted samples.
- The existing code in SpeziHealthKit has been changed to use these new APIs where possible
- Note that we currently intentionally only have one-off querying APIs; we could also implement something that 


## :gear: Release Notes
- Added several `SampleType<HKCategorySample>` extensions, for medical symptoms
- Added some new methods on the HealthKit module, which allow for type-safe querying of HealthKit data outside of SwiftUI views:
    - `query(_:timeRange:limit:predicate:)`
    - `anchorQuery(_:timeRange:anchor:limit:predicate:)`
    - `continuousQuery(_:startTime:anchor:limit:predicate:)`


## :books: Documentation
The new code is documented


## :white_check_mark: Testing
The new code is implicitly tested, since it is used as part of the other, existing functionality.


## :pencil: Code of Conduct & Contributing Guidelines
By creating and submitting this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

PSchmiedmayer commented on Sources/SpeziHealthKit/Queries/HealthKitQuery.swift:
```diff
@@ -0,0 +1,150 @@
+    /// // This example fetches all heart rate samples
+    /// // of the past half year, 5000 samples at a time
+    /// var anchor = QueryAnchor<HKQuantitySample>()
+    /// while true {
```
Not sure if a while `true` loop example is a great demonstration of the functionality, maybe rather add an example that stores the anchor somewhere in a type and offers a functionality to query the next samples or something similar?

lukaskollmer replied:
i thought about that as well, but the reason i went with this example specifically was that:
1. we don't have a dedicated batch query function, and this demonstrates that it's easy use the existing functions to get batching behaviour
2. it demonstrates that the anchors are lightweight objects which can just be created and used on-demand for something like this. IMP that also implies that you can store them and re-use them later on.

tbh, i do agree that we should have an example with that as well, maybe in the documentation of the `QueryAnchor` type itself?

PSchmiedmayer replied:
Sounds good; we can keep the example as it is 👍

Might be good to add a nice AsyncSequence for this use case in the future.

PSchmiedmayer commented on Sources/SpeziHealthKit/Queries/HealthKitQuery.swift:
```diff
@@ -0,0 +1,150 @@
+        anchor: QueryAnchor<Sample>,
+        limit: Int? = nil,
+        predicate filterPredicate: NSPredicate? = nil
+    ) -> AsyncMapSequence<HKAnchoredObjectQueryDescriptor<Sample>.Results, ContinuousQueryElement<Sample>> {
+        // ^^^ we sadly can't use `some AsyncSequence<ContinuousQueryElement<Sample>, any Error>` here,
+        // since the Failure parameter is only available starting with iOS 18...
```
What about making this function iOS18 only?

Or if that's not feasible, at least have an overload that is available for older versions but deprecated for iOS 18 and newer and then points to the function signature that has a proper failure parameter in place? We will have to make this a breaking release anyways.

PSchmiedmayer replied:
Sounds great; I would aim for the overload approach and pushing people to the better version for iOS 18.

lukaskollmer replied:
update: we can do this with an overload (though we'd need 2 additional functions to make this work).

to be honest, in my opinion having the explicit type instead of the `some` type is fine here. it behaves the same way as the some type, and it keeps our implementaton a lot cleaner than doing the overload stuff

lukaskollmer replied:
@PSchmiedmayer i implemented and pushed the overload-based workaround. let me know what you think

PSchmiedmayer commented on Sources/SpeziHealthKit/Sample Types/SampleTypes.swift.gyb:
```diff
@@ -889,6 +889,213 @@ category_types = [
         doc='A category sample type for handwashing events.'
+    ),
+
+    # Symptoms
```
Thank you for adding all the symptoms here!

PSchmiedmayer commented on Sources/SpeziHealthKit/SpeziHealthKit.docc/SpeziHealthKit.md:
Seems like there is a link checking error here: `Cannot reach doc:ModuleConfiguration. Status: null net::ERR_ABORTED at doc:ModuleConfiguration`

lukaskollmer replied:
it seems that the markdown checker is treating the `<doc:ModuleConfiguration>` link as if it were a normal, web-accessible link. i'm not sure what we can do about this.

PSchmiedmayer reviewed (approved):
Looks solid, thank you for all the improvements!

lukaskollmer commented on Sources/SpeziHealthKit/Queries/QueryAnchor.swift:
```diff
@@ -0,0 +1,50 @@
+
+
+/// Used by anchor queries to keep track of the last-seen state of the HealthKit database.
+public struct QueryAnchor<Sample: _HKSampleWithSampleType>: Hashable, Codable, Sendable {
```
@PSchmiedmayer i'd appreciate your feedback on this, and whether you think we should keep this type, or just use a plain `HKQueryAnchor` instead.

my idea here was to make the anchors directly Codable, so that users can store (and re-use) them using a system  of their choosing.
i also figured that making the anchor generic over a sample type would make it more difficult to accidentally use the wrong anchor somewhere (i'm not really sure whether this part was actually a good idea, tbh)

there's three paths here:
1. keep it as-is
2. keep it but make it non-generic
3. just use a direct `HKQueryAnchor` instead

what's your take?

PSchmiedmayer replied:
I would vote for option 2.

lukaskollmer commented:
@PSchmiedmayer i discovered some more missing sample types and added them as well (it's annoying that Apple doesn't have an official list where they just have all of them in one place; but i think we probably should have all of them now). i also added dedicated macOS and watchOS steps to the GH actions, since we do support these platforms, and there's some differences in terms of which HealthKit types are available on which platforms
